### PR TITLE
Add SyTest support to CI

### PR DESCRIPTION
**Pull request by anoadragon453**
Adds SyTest CI through CircleCI.

Ideally https://github.com/matrix-org/sytest/pull/604 will be merged as well which will give us more meaningful TAP output.

Also need to add all the tests that currently pass to `testfile`, but I don't think that's a hard blocker for merging.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+          name: Build Dendrite
+          command: ./build.sh
+      - run: 
+          name: Copy custom dendrite config
+          command: cp .circleci/dendrite-config.yaml dendrite.yaml
+      - run: 
+          name: Generating self-signed server keys
+          command: |
+            test -f server.key || openssl req -x509 -newkey rsa:4096 -keyout server.key -out server.crt -days 3650 -nodes -subj /CN=localhost
+            test -f matrix_key.pem || ./bin/generate-keys -private-key matrix_key.pem
```
Comment (erikjohnston) [inline .circleci/config.yml]: Can't we just keep a static set in `.circleci/` so that we don't have to wait for them to be generated each run?

**Comment by babolivier** [comment]
> I vote we merge and then we can actually test what happens when the circleci gets run

I don't know CircleCI enough, but can't we enable it for the repo anyway, let it choke everywhere except for this branch where it actually has a config?

**Comment by erikjohnston** [comment]
You can try, but I couldn't figure out how.

**Comment by anoadragon453** [comment]
Can view how it works here: https://github.com/anoadragon453/dendrite/commits/master

Breaks on latest commit as I need to update the matrixdotorg docker images, but waiting for https://github.com/matrix-org/sytest/pull/611 to be merged first.

**Comment by babolivier** [comment]
> You can try, but I couldn't figure out how.

I think I've done it, let's wait till another commit arrives to check whether it works.

**Comment by anoadragon453** [comment]
Seems to have worked.

File: testfile [hunk 2]
```diff
@@ -0,0 +1,3 @@
+GET /register yields a set of flows
+POST /register can create a user
+POST /register downcases capitals in usernames
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply testfile]: > Also need to add all the tests that currently pass to testfile, but I don't think that's a hard blocker for merging.

Was waiting for that one PR to merge before trudging through the list, so I can do it now I suppose.

File: .circleci/dendrite-config.yaml [hunk 3]
```diff
@@ -0,0 +1,147 @@
+# The config file format version
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply .circleci/dendrite-config.yaml]: Probs.

File: testfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply testfile]: IRL: we're doing it later.

**Pull request merged**